### PR TITLE
SOT-123: Write Fast Fourier Transforms

### PR DESCRIPTION
**Pull request by VanhGer**
# Check List

- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!
- [x] Did you verify that your math formulas are rendered correctly?

### CODE REVIEW

File: terms/fast_fourier_transforms.md [hunk 1]
```diff
@@ -1,3 +1,157 @@
...
+To use this algorithm, you need to use finite field, and the domain must be a [multiplicative subgroup](./prime_or_finite_fields) of
+the field. Furthermore, the multiplicative subgroup must have size $2^n$ in order to create a low-runtime recursive algorithm.
+
+**High level idea:**
+FFT works like [merge-sort](https://www.geeksforgeeks.org/merge-sort/) . It takes a problem of size $N$, break it up into two problems
+of size $\dfrac{N}{2}$, and do $O(N)$ work to combine two smaller solutions into the bigger one. So, the total run time is $O(NlogN)$.
+
+### Multi-point evaluation
+
+Suppose that we have an FFT where the given domain is the powers of $w$ in some field, where $w^{2^k} = 1$ ($w$ is the domain
```
Comment (hduoc2003) [inline terms/fast_fourier_transforms.md:40]: math problem

File: terms/fast_fourier_transforms.md [hunk 2]
```diff
@@ -1,3 +1,152 @@
...
+
+Hence, we can use FFT to compute $evens(x)$ or $odds(x)$ for every $x$ in the smaller domain consisting of squares of numbers in the
+original domain. And, we've reduced a size-$N$ problem into half-size problem.
+
+After receive the evaluation of $evens$ and $odds$, we simply do:
+
+- $f[i] = evens[i] + domain[i] \cdot odds[i]$
+- $f[N/2 + i] = evens[i] - domain[i] \cdot odds[i]$
+
+for each index $i$
```
Comment (hduoc2003) [inline terms/fast_fourier_transforms.md:73]: $i$ from $0$ to $N/2-1$

File: terms/fast_fourier_transforms.md [hunk 3]
```diff
@@ -1,3 +1,152 @@
...
+
+Python code for this:
+
+```
+def fft(vals, modulus, domain):
```
Comment (hduoc2003) [inline terms/fast_fourier_transforms.md:78]: add python annotation for this code block
image

File: terms/prime_or_finite_fields.md [hunk 4]
```diff
@@ -13,6 +13,21 @@ defined as follows:
...
+
+For a informal explanation, $G$ is a list of powers of some "generator" value.
+
+For example, in the finite field $F_{17}$, if we choose the generator is
+$4$,  $G = \lbrace 1, 4, 16, 13 \rbrace$:
+
+- $4^0 = 1$
+- $4^1 = 4$
+- $4^2=16$
+- $4^3= 64 \equiv 13$ (mod 17)
```
Comment (hduoc2003) [inline terms/prime_or_finite_fields.md:26]: \pmod

File: terms/fast_fourier_transforms.md [hunk 5]
```diff
@@ -1,3 +1,152 @@
...
+- $f(x) = evens(x^2) + x \cdot odds(x^2)$
+- $f(-x) = evens(x^2) - x \cdot odds(x^2)$
+
+Here, both $evens$ and $odds$ are polynomials half the size of $f(x)$ and the set of possible values of $x^2$ is only half the size of
+the original domain: \[1, 4, 16, 13\] $\to$ \[1, 16\]. The reason is that:
+
+- $1^2 = 1$
+- $4^2 = 16$
+- $16^2 = 256 \equiv 1\mod 17$
+- $13^2 = 169 \equiv 16 \mod 17$
```
Comment (hduoc2003) [inline terms/fast_fourier_transforms.md:63]: for more concrete, $1 \equiv \omega^{|G|}$, so $(\omega^k)^2 \equiv (\omega^{k+|G|/2})^2$

File: terms/fast_fourier_transforms.md [hunk 6]
```diff
@@ -1,3 +1,153 @@
...
 
-[Link](https://vitalik.eth.limo/general/2019/05/12/fft.html)
+## Background
+
+First, you should have a basic understanding of Fourier Transforms. Essentially,  Fourier Transforms convert data between the
+"frequency domain" and the "time domain". It means that if you have a pieces of data, the algorithm will transform it into a
+collection of sine waves with different frequencies and amplitudes. Then, adding them together would approximate the
+original data. See the depiction below.
+
+![](attachments/fourier_transforms.png)
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:10]: you should consider adding the name of each figure

File: terms/prime_or_finite_fields.md [hunk 7]
```diff
@@ -13,6 +13,21 @@ defined as follows:
 
 In essence, all mathematical operations are performed modulo $p$.
 
+### Multiplicative Group
+
+For a informal explanation, $G$ is a list of powers of some "generator" value.
```
Comment (bjergsen243) [inline terms/prime_or_finite_fields.md:18]: For an informal

File: terms/prime_or_finite_fields.md [hunk 8]
```diff
@@ -13,6 +13,21 @@ defined as follows:
...
+
+For example, in the finite field $F_{17}$, if we choose the generator is
+$4$,  $G = \lbrace 1, 4, 16, 13 \rbrace$:
```
Comment (bjergsen243) [inline terms/prime_or_finite_fields.md:21]: if ..., then G = .... Because of ...

File: terms/prime_or_finite_fields.md [hunk 9]
```diff
@@ -13,6 +13,21 @@ defined as follows:
...
+
+- $4^0 = 1$
+- $4^1 = 4$
+- $4^2=16$
+- $4^3= 64 \equiv 13$ (mod 17)
+
+Multiplicate group is cyclic. The prove can be found
```
Comment (bjergsen243) [inline terms/prime_or_finite_fields.md:28]: The proof

File: terms/fast_fourier_transforms.md [hunk 10]
```diff
@@ -1,3 +1,153 @@
...
+Then, do the same things as in the FFT section.
+
+Here is the code for all the implementation that we mentioned above:
+[link](https://github.com/ethereum/research/tree/master/binary_fft).
+
+## Reference
+
+[Fast Fourier Transforms](https://vitalik.eth.limo/general/2019/05/12/fft.html)
+<br>
+Vitalik Buterin
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:153]: If Vitalik's article was already linked above, there's no need to mention his name again

File: terms/fast_fourier_transforms.md [hunk 11]
```diff
@@ -1,3 +1,153 @@
 # Fast Fourier Transforms
 
-[Link](https://vitalik.eth.limo/general/2019/05/12/fft.html)
+## Background
+
+First, you should have a basic understanding of Fourier Transforms. Essentially,  Fourier Transforms convert data between the
```
Comment (sonntuet1997) [inline terms/fast_fourier_transforms.md:5]:
```suggestion
Essentially,  Fourier Transforms convert data between the
```

File: terms/fast_fourier_transforms.md [hunk 12]
```diff
@@ -1,3 +1,153 @@
...
+"frequency domain" and the "time domain". It means that if you have a pieces of data, the algorithm will transform it into a
```
Comment (sonntuet1997) [inline terms/fast_fourier_transforms.md:6]:
```suggestion
"frequency domain" and the "time domain". It means that if you have a piece of data, the algorithm will transform it into a
```

File: terms/fast_fourier_transforms.md [hunk 13]
```diff
@@ -1,3 +1,153 @@
...
+
+The Fourier Transforms we discuss are discrete Fourier Transforms over the [finite field](./prime_or_finite_fields.md), rather than the
+continuous Fourier Transforms over the *real or complex number* above. And, instead of focusing on the conversion between
+"frequency domain" and "time domain", we primarily talk about two operations: **multi-point polynomial evaluation** and its
+inverse, **polynomial interpolation**.
+
+First, **multi-point polynomial evaluation** involves evaluating a degree $\lt N$ polynomial at $N$ different points, and
+**polynomial interpolation** is recovering the polynomial using the given evaluation of its at $N$ different points.
+
+For the first operation, you all can separately evaluate the polynomial at each point just by iterating over all the coefficients.
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:20]:
```suggestion
For the first operation, you can evaluate the polynomial at each point by iterating over all the coefficients separately
```

File: terms/fast_fourier_transforms.md [hunk 14]
```diff
@@ -1,3 +1,153 @@
...
+For the second operation, you first need to read about [Lagrange interpolation](./lagrange_interpolation.md) to understand how it
+works and what Lagrange basis polynomials are:
+$$L_i(x) = \prod_{j=0, j \neq i}^{n} \frac{x - x_j}{x_i - x_j}$$
+Compute this function takes $O(N^2)$ time (iterate $j$ take $O(N)$ times, and we need to compute $N$ functions).
+
+Is this slow? Yes, and the Fast Fourier Transforms below will significantly speed up both operations.
+
+## Fast Fourier Transforms
+
+To use this algorithm, you need to use finite field, and the domain must be a [multiplicative subgroup](./prime_or_finite_fields) of
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:31]:
```suggestion
To utilize this algorithm, you must operate within a finite field, and the domain must be a multiplicative subgroup of that field. Additionally, the size of the multiplicative subgroup must be [size] to enable the creation of a low-runtime recursive algorithm
```

File: terms/fast_fourier_transforms.md [hunk 15]
```diff
@@ -1,3 +1,153 @@
...
+        return vals
+    L = fft(vals[::2], modulus, domain[::2])
+    R = fft(vals[1::2], modulus, domain[::2])
+    o = [0 for i in vals]
+    for i, (x, y) in enumerate(zip(L, R)):
+        y_times_root = y*domain[i]
+        o[i] = (x+y_times_root) % modulus
+        o[i+len(L)] = (x-y_times_root) % modulus
+    return o
+```
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:90]: do you have input example to test?

File: terms/fast_fourier_transforms.md [hunk 16]
```diff
@@ -1,3 +1,153 @@
...
+
+### Polynomial Interpolation
+
+To compute inverse FFT, we just run the FFT again , but reverse the result (except the first item stays in place) and divide every
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:94]: remove 
```suggestion
To compute the inverse FFT, we run the FFT again, but reverse the result (keeping the first item in place) and divide every term by the length of the input sequence.
```

File: terms/fast_fourier_transforms.md [hunk 17]
```diff
@@ -1,3 +1,153 @@
...
+multiplication.
+
+## FFTs and Binary Fields
+
+In an binary field, each element is expressed as a polynomial where all of the entries are 0 or 1, eg. $x^3+x+1$. Adding polynomials
+is done modulo 2, and subtraction is the same as addition (as $−1=1 \mod2$). We select a irreducible polynomial as a modulus
+(e.g., $x^4 + x + 1$) and multiplication is done modulo that modulus.
+
+If you want to run FFT in a binary field, there is a problem:
+In binary fields, there are *no multiplicative groups of order $2^n$* because all the multiplicative groups are order $2^n - 1$.
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:114]: fix math display error at 2^n

File: terms/fast_fourier_transforms.md [hunk 18]
```diff
@@ -1,3 +1,153 @@
...
+    vals = fft(vals, modulus, domain)
+    return [x * modular_inverse(len(vals), modulus) % modulus for x in [vals[0]] + vals[1:][::-1]]
+```
+
+We can use FFT to multiply numbers in time $O(NlogN)$ which $N$ is the length of numbers, faster than $O(N^2)$ schoolbook
+multiplication.
+
+## FFTs and Binary Fields
+
+In an binary field, each element is expressed as a polynomial where all of the entries are 0 or 1, eg. $x^3+x+1$. Adding polynomials
```
Comment (bjergsen243) [inline terms/fast_fourier_transforms.md:109]: a binary field

File: terms/fast_fourier_transforms.md [hunk 19]
```diff
@@ -1,3 +1,153 @@
...
+original data. See the depiction below.
+
+![](attachments/fourier_transforms.png)
+
+The Fourier Transforms we discuss are discrete Fourier Transforms over the [finite field](./prime_or_finite_fields.md), rather than the
+continuous Fourier Transforms over the *real or complex number* above. And, instead of focusing on the conversion between
+"frequency domain" and "time domain", we primarily talk about two operations: **multi-point polynomial evaluation** and its
+inverse, **polynomial interpolation**.
+
+First, **multi-point polynomial evaluation** involves evaluating a degree $\lt N$ polynomial at $N$ different points, and
```
Comment (sonntuet1997) [inline terms/fast_fourier_transforms.md:17]: Rewrite this

File: terms/fast_fourier_transforms.md [hunk 20]
```diff
@@ -1,3 +1,153 @@
...
+
+- $1^2 = 1$
+- $4^2 = 16$
+- $16^2 = 256 \equiv 1\pmod {17}$
+- $13^2 = 169 \equiv 16 \pmod {17}$
+
+Hence, we can use FFT to compute $evens(x)$ or $odds(x)$ for every $x$ in the smaller domain consisting of squares of numbers in the
+original domain. And, we've reduced a size-$N$ problem into half-size problem.
+
+After receive the evaluation of $evens$ and $odds$, we simply do:
```
Comment (sonntuet1997) [inline terms/fast_fourier_transforms.md:68]:
```suggestion

After receiving the evaluation of $evens$ and $odds$, we simply do:
```

File: terms/fast_fourier_transforms.md [hunk 21]
```diff
@@ -1,3 +1,153 @@
...
+    return o
+```
+
+### Polynomial Interpolation
+
+To compute inverse FFT, we just run the FFT again , but reverse the result (except the first item stays in place) and divide every
+value by the length of the list.
+An inverse FFT is surprisingly simple:
+
+```python
```
Comment (sonntuet1997) [inline terms/fast_fourier_transforms.md:98]: Screenshot 2024-03-23 at 16 32 40
https://www.youtube.com/watch?v=h7apO7q16V0

Comment (sonntuet1997) [reply terms/fast_fourier_transforms.md:98]: Screenshot 2024-03-23 at 16 32 14

**Review by sonntuet1997 — APPROVED** [review]
Fix the comments and it's good to go

File: terms/prime_or_finite_fields.md (hunk 8 above)
Comment (ktokto313) [inline terms/prime_or_finite_fields.md:21]: I think it should be "if we choose the generator 4" or "if we choose 4 as the generator"

**Review by ktokto313 — APPROVED** [review]
Look good ..

**Review by bjergsen243 — APPROVED** [review]
lgtm

**Pull request merged**